README.md
Poker-Bot-Reformed
==================

Not finished yet.
---


CHANGELOG
-----------------------------
VERSION 3.0

x 写一个简单的学习机器人，Preflop按照规则进行动作，Postflop则按board与line分成不同的情况，记录不同情况
  下不同动作的导致的结果，作为未来行动的依据

-----------------------------

阶段总结
在VERSION 0~2中，核心的思路都是计算equity，并根据equity来估算自己的不同行为会带来多少收益以及对手持不同
牌做不同行为的概率。本质上来说，就是设定了许许多多的规则。但是情况是复杂的，不同的对手在不同的牌面上做
动作的概率分布是不一样的，如果要设定规则，那就需要设定很详细具体的规则，这是一件很麻烦的事情。最致命的
是，如果想让程序不断进步，那么就需要不断地人工维护这些规则，这显然是做不到的。
经过一些实验，如果精心调试，VERSION 2系列的程序应该可以在NL2中获得胜利，NL5也应该问题不大。不过，精心调
试会消耗大量的时间，而且程序不具有进步的能力，因此放弃。在以后的版本中，具有学习能力是必要条件。

-----------------------------
VERSION 2.1

o 现在按照dummy_action来更新对手的stats，效果还不错
o 提速

x 重写strategy
x 为每个人都搞一个win chance table，而不是用平均数据来替代

-----------------------------
VERSION 2.0

o 改bug，美化输出
o 调整对stats改变的算法

-----------------------------
VERSION 1.1

o 加入了Simulate_Value功能，从而可以智能做出决定以及判断对手动作的概率。不过从实践上来看效果不好。
o 人工规定了一个dummy，假设对手进行动作时的对象是dummy，效果一般。

T 大约6000手，EV-10

-----------------------------
VERSION 1.0

o 重新改写了win chance的计算，现在使用的是真正的win chance而不是通过power rank和outs table得到的近似win chance
o 增加了show win chance的功能

-----------------------------
VERSION 0.6

o 更改胜率计算，现在胜率=不败概率

-----------------------------
VERSION 0.5

o 有showdown价值的中等牌力的牌不要后位开枪了

-----------------------------
VERSION 0.4

o 在脏面上多下注
o steal多一些

T 2014.10.3 HANDS:1000 NET:-3

-----------------------------
VERSION 0.3

o River raise小一些
o 加入了board_wetness
o 拿到nuts的时候在干燥面Flop前位可以考虑check call
o 在3bet后谨慎bluff
o 在bluff的时候考虑range_fight_result
o 用更多的draw买
o 在后位的时候用更多的牌来open
o 修正了偶尔出现的莫名其妙卡住并且无法自动恢复的bug

T 2014.10.1~2014.10.2 HANDS:3500 NET:-1.6

------------------------------
VERSION 0.2

o 拿着nuts的时候装作思考，暂停一小会再下注
o 修正卡住至inactive的bug
o Preflop的超大额加注算作多次Bet
o 改正new game的判定方法，如果手牌变了就是new game

T 2014.9.29~2014.9.30 HANDS:3200 NET:0.x

------------------------------
VERSION 0.1

o 计算beat_chance时，不考虑对手概率小于0.1的组合
o 计算outs时，高张的outs不按1来算，为outs的强弱赋予不同的数值
o 计算ratio时，考虑自己的后手筹码量
o error analysis时对手牌进行shuffle

------------------------------
VERSION 0.0

o 框架完成，可以长时间无缝运行
o 基本思路是通过range判断自己的胜率，如果胜率较高则下注，否则过牌弃牌
o 二人局时，Turn\Flop后位必开枪
o 二人局非3Bet时，Flop必CB


game_driver.py
#coding: utf-8

import json
import re
import time
from public import is_only_max
from public import get_win_chance_table 
from public import show_stats
from public import change_terminal_color
from pokerstars.config import BB, SB
from pokerstars.screen_scraper import ScreenScraper
from pokerstars.move_catcher import MoveCatcher
from pokerstars.controller import Controller
from database.data_manager import DataManager
from stats.stats_handler3 import StatsHandler
from strategy.decision_maker2 import PreflopDecisionMaker
from strategy.decision_maker2 import PostflopDecisionMaker

class GameDriver():

    def __init__(self, game_record='ps', shift=[0, 0]):
        self.screen_scraper = ScreenScraper(game_driver=self, source=game_record, shift=shift)#{{{
        self.shift          = shift
        self.source         = game_record
        self.controller     = Controller(self, shift)
        init_values = self.screen_scraper.get_init_values()
        self.new_game_indicator = 0
        if init_values in ['sit out', 'get back']:
            if init_values == 'sit out':
                self.controller.sit_out()
            else:
                self.controller.get_back()
            self.new_game_indicator = 1
        if self.new_game_indicator == 0:
            self.stack          = init_values['stack'] 
            self.game_number    = init_values['game_number']
            self.cards          = init_values['cards'] 
            self.button         = init_values['button']
            self.player_name = init_values['player_name']
            if self.source != 'ps':
                self.seat       = init_values['seat']
            self.steal_position = self.button == 0 or self.button == 1
            self.active         = [1, 1, 1, 1, 1, 1]
            self.data_manager   = DataManager(self.button)
            if self.source != 'ps':
                self.player_name = map(lambda x:unicode(x), self.player_name)
            self.data_manager.load_data(self.player_name, self.button)
            self.stats_handler  = StatsHandler(self)
            self.decision_maker = PreflopDecisionMaker(self)
            self.betting        = [0, 0, 0, 0, 0, 0]
            self.pot            = SB+BB
            self.last_better    = -1
            self.all_limper     = -1
            self.stage          = 0
            self.bet_round      = 1
            self.people_play    = 1
            self.move_catcher   = MoveCatcher(0, self)
            self.postflop_status = ['', '', '', '', '', '']
            self.power_rank     = [0, 0, 0, 0]
            self.can_beat_table = [0, 0, 0, 0]
            self.board_wetness  = [0, 0, 0, 0]
            self.fold           = 0
            self.outs           = [0, 0, 0, 0]#}}}

    @classmethod
    def count_game(cls):
        try:#{{{
            cls.game_count += 1 
        except:
            cls.game_count = 1#}}}

    def game_stream(self, last_game):
        if self.new_game_indicator == 1:
            return 0
        print 'Game Number:', self.game_number
        if self.game_number == last_game:
            change_terminal_color('green')
            print 'game over because game_number == last_game'
            change_terminal_color()
            return last_game
        print 'Stack:', self.stack#{{{
        print 'Button:', self.button
        print 'Cards:', self.cards[:2]
        print 'Names:', self.player_name
        for i in xrange(1, 6):
            if type(self.player_name[i]) == unicode:
                print '%s, Hands: %d, PFR: %0.2f, F3B: %0.2f, BSA: %0.2f' %\
                        (self.player_name[i], self.data_manager.get_item(i, u'HANDS'),\
                        self.data_manager.get_item(i, u'pfr'), self.data_manager.get_item(i, u'F3B'),\
                        self.data_manager.get_item(i, u'BSA'))
        if self.stack[0] > 400*SB and self.decision_maker.get_preflop_move(self.cards) == 0:
            self.controller.sit_out()
            return self.game_number
        if self.source == 'ps' and self.decision_maker.get_preflop_move(self.cards) == 0 and self.button != 4 and self.button != 0:
            change_terminal_color('green')
            print 'Fold because my preflop move is 0'
            change_terminal_color()
            self.controller.fold()
            return self.game_number
        change_terminal_color()
        print#}}}
        indicator = self.preflop()#{{{
        self.decision_maker = PostflopDecisionMaker(self, self.stats_handler.stats)
        stages = ['PREFLOP', 'FLOP', 'TURN', 'RIVER']
        self.stage = 1
        if indicator == 'new game':
            change_terminal_color('green')
            print 'game over because preflop return so'
            change_terminal_color()
            return self.game_number
        for self.stage in xrange(1, 4):
            change_terminal_color('blue')
            print '*** '+stages[self.stage]+' ***'
            print 'Pot: ', self.pot
            if self.stage == 1:
                print 'Cards: ', self.cards[2:5]
            elif self.stage == 2:
                print 'Cards: ', self.cards[2:5], self.cards[5]
            elif self.stage == 3:
                print 'Cards: ', self.cards[2:5], self.cards[5], self.cards[6]
            change_terminal_color()
            indicator = self.post_flop(self.stage)
            if indicator == 'new game':
                return self.game_number#}}}
    
    def handle_preflop_action(self, action):
        if action[0] == 'new game':#{{{
#           print 'new game'
            return 'new game'
        if action[0] == 'new stage':
#           print 'new stage'
            self.cards = action[1]
            return 'new stage'
        if action[0] == 'my move':
#           print 'making decision'
            self.decision_maker.make_decision(self)
            time.sleep(0.5)
            return []
        actor, value = action
        if value == 'fold':
            print 'Player '+str(actor)+': Fold'+' '*12+'\t <-- '+str(self.betting)
            self.active[actor] = 0
            return []
        if value == 'check':
            print 'Player '+str(actor)+': Check'+' '*11+'\t <-- '+str(self.betting)
            return []
#       self.betting[actor] = round(self.betting[actor]+value, 2)
        if is_only_max(self.betting, actor):
            print 'Player '+str(actor)+': Raise --> '+str(self.betting[actor])+'  \t <-- '+str(self.betting)
        else:
            print 'Player '+str(actor)+': Call'+' '*12+'\t <-- '+str(self.betting)
        if is_only_max(self.betting, actor):
            self.last_better = actor
            self.bet_round += 1
            if self.bet_round == 2:
                if self.betting[actor] / BB > 5 + self.people_play:
                    self.bet_round += 1
            if self.bet_round == 3:
                if self.betting[actor] > 3.5 * self.pot:
                    self.bet_round += 1
            self.people_play = 1
        else:
            if self.last_better == -1:
                self.last_better = actor
            self.people_play += 1
        self.stats_handler.preflop_update(action, self.betting, self.bet_round,\
                self.people_play, self.last_better)
#        if self.source != 'ps':
#            show_stats(self.stats_handler.stats, actor)
        self.pot += value
        self.pot = round(self.pot, 2)
        self.stack[actor] -= value
        self.stack[actor] = round(self.stack[actor], 2)
        if self.stack[actor] == 0:
            self.active[actor] = 0.5
        return []#}}}

    def handle_postflop_action(self, action):
        if action[0] == 'new game':#{{{
#           print 'new game'
            return 'new game'
        if action[0] == 'new stage':
#           print 'new stage'
            self.cards = action[1]
            return 'new stage'
        if action[0] == self.last_mover and action[1] == 'check' and self.stage == 3:
            print 'stage', self.stage
            return 'new game'
        if action[0] == 'my move':
            if self.stage == 3:
                a = self.betting[0]
                not_even = 0
                for i in xrange(6):
                    if self.active[i] == 1 and self.betting[i] != a:
                        not_even = 1
            if self.stage != 3 or not_even or a == 0 or (a < max(self.betting) and self.active[0] == 1):
                self.decision_maker.make_decision()
                time.sleep(0.5)
            return []
        actor, value = action
        if value == 'fold':
            print 'Player '+str(actor)+': Fold'+' '*12+'\t <-- '+str(self.betting)
            self.active[actor] = 0
            return []
        if value == 'check':
            print 'Player '+str(actor)+': Check'+' '*11+'\t <-- '+str(self.betting)
            self.postflop_status[actor] = 'check'
            value = 0
        else:
#           self.betting[actor] = round(self.betting[actor]+value, 2)
            if is_only_max(self.betting, actor):
                print 'Player '+str(actor)+': Raise --> '+str(self.betting[actor])+'  \t <-- '+str(self.betting)
            else:
                print 'Player '+str(actor)+': Call'+' '*12+'\t <-- '+str(self.betting)
            self.stack[actor] -= value
            self.stack[actor] = round(self.stack[actor], 2)
            if self.stack[actor] == 0:
                self.active[actor] = 0.5
            if is_only_max(self.betting, actor):
                if max(self.betting) == sum(self.betting): 
                    if max(self.betting) < self.pot*0.2 and self.stage != 3:
                        self.postflop_status[actor] = 'check'
                    elif self.last_better == actor:
                        self.postflop_status[actor] = 'bet'
                    else:
                        self.postflop_status[actor] = 'bet'
                else:
                    if self.postflop_status[self.last_better] in ['raise', 'reraise', 'check raise']:
                        self.postflop_status[actor] = 'reraise'
                    elif self.postflop_status[self.last_better] == 'check':
                        self.postflop_status[actor] = 'bet'
                    else:
                        if self.postflop_status[actor] == 'check':
                            self.postflop_status[actor] = 'check raise'
                        else:
                            self.postflop_status[actor] = 'raise'
                self.last_better = actor
            else:
                if self.postflop_status[actor] == 'check':
                    self.postflop_status[actor] = 'check call'
                else:
                    self.postflop_status[actor] = 'call '+\
                            self.postflop_status[self.last_better]
            self.pot += value
            self.pot = round(self.pot, 2)
        if sum([self.active[i] == 1 for i in xrange(6)]) <= 1\
                and (self.active[0] != 1 or self.betting[0] == max(self.betting)):
            print 'stage', self.stage
            print 'active', self.active
            return 'new game'
        all_even = 1
        for i in xrange(6):
            if not self.active[i] == 1:
                continue
            if self.betting[i] != max(self.betting):
                all_even = 0
                break
        if all_even and max(self.betting) > 0 and self.stage == 3:
            print 'stage', self.stage
            print 'active', self.active
            return 'new game'
        self.decision_maker.update_status()
        self.decision_maker.update_stats(actor, value)
        self.decision_maker.get_avg_stats()
        self.decision_maker.get_opponent()
        self.decision_maker.update_betting()
        return []#}}}

    def preflop(self):
        to_act = (self.button+3) % 6#{{{
        self.betting = [0, 0, 0, 0, 0, 0]
        self.betting[(self.button+1)%6] = SB
        self.betting[(self.button+2)%6] = BB
        self.people_play = 1
        self.last_mover = (self.button+2) % 6
        self.move_catcher.to_act = to_act
#       move_catcher = MoveCatcher(to_act, self)#}}}
        while True:
#            if sum([self.active[j] == 1 for j in xrange(1, 6)]) == 0:
#                change_terminal_color('green')
#                print 'game over because no active opponent'
#                change_terminal_color()
#                return 'new game'
            actions = self.move_catcher.get_action()
            if self.source == 'ps':#{{{
                for action in actions:
                    if action[0] == 'new game':
                        change_terminal_color('green')
                        print 'game over because move catcher say so'
                        change_terminal_color()
                        return 'new game' 
                if not actions:
                    if self.button == 0 and\
                            (type(self.player_name[1]) != unicode\
                            or type(self.player_name[2]) != unicode):
                        print 'Wait, Because: ', self.player_name[1:3]
                    elif self.button == 4 and\
                            (self.data_manager.get_item(4, u'BSA') > 0.7):
                        pass
                    else:
                        self.decision_maker.fast_fold(self)
                for action in actions:
                    if action[0] in [1,2,3,4,5] and self.betting[0] < max(self.betting):
                        if self.button == 0 and\
                                (type(self.player_name[1]) != unicode\
                                or type(self.player_name[2]) != unicode):
                            print 'Wait, Because: ', self.player_name[1:3]
                        elif self.button == 4 and\
                                (self.data_manager.get_item(4, u'BSA') > 0.7):
                            pass
                        else:
                            self.decision_maker.fast_fold(self)
                    indicator = self.handle_preflop_action(action)
                    if indicator:
                        change_terminal_color('green')
                        print 'game over because handle_preflop_action say so'
                        change_terminal_color()
                        return indicator
                if sum([self.active[i] == 1 for i in xrange(6)]) <= 1\
                        and (self.active[0] != 1 or self.betting[0] == max(self.betting)):
                    change_terminal_color('green')
                    print 'game over because there is no active player'
                    change_terminal_color()
                    return 'new game'
                to_act = self.move_catcher.to_act#}}}
            else:#{{{
                action = actions[0]
                indicator = self.handle_preflop_action(action)
                if indicator:
                    return indicator
                to_act = self.move_catcher.to_act#}}}
            
    def post_flop(self, stage):
        self.postflop_status = ['', '', '', '', '', '']#{{{
        to_act = (self.button+1) % 6
        self.betting = [0, 0, 0, 0, 0, 0]
        self.last_mover = self.button
        while self.active[self.last_mover] != 1:
            self.last_mover = (self.last_mover-1) % 6
        self.move_catcher.to_act = to_act
        self.decision_maker.update_status()
        self.decision_maker.clean_stats()
        self.decision_maker.compress_stats()
        self.decision_maker.get_avg_stats()
        self.decision_maker.get_opponent()
        self.decision_maker.update_wct()
        self.decision_maker.get_dummy_action_table()
#       move_catcher = MoveCatcher(to_act, self)#}}}
        while True:
#            if sum([self.active[j] == 1 for j in xrange(6)]) <= 1:
#                return 'new game'
            actions = self.move_catcher.get_action()
            if self.source == 'ps':#{{{
                for action in actions:
                    if action[0] == 'new game':
                        return 'new game' 
                for action in actions:
                    if action[0] in [1,2,3,4,5] and self.betting[0] < max(self.betting):
                        self.decision_maker.fast_fold()
                    indicator = self.handle_postflop_action(action)
                    if indicator:
                        return indicator
                if sum([self.active[i] == 1 for i in xrange(6)]) <= 1\
                        and (self.active[0] != 1 or self.betting[0] == max(self.betting)):
                    change_terminal_color('green')
                    print 'game over because there is no active player'
                    change_terminal_color()
                    return 'new game'
                to_act = self.move_catcher.to_act#}}}
            else:#{{{
                if len(actions) == 1:
                    action = actions[0]
                    indicator = self.handle_postflop_action(action)
                    if indicator:
                        return indicator
                else:
                    action = actions[1]
                    if type(action[1]) == float:
                        self.betting[0] -= action[1]
                        self.betting[0] = round(self.betting[0], 2)
                    action = actions[0]
                    indicator = self.handle_postflop_action(action)
                    action = actions[1]
                    self.handle_postflop_action(action)
                    if type(action[1]) == float:
                        self.betting[0] += action[1]
                        self.betting[0] = round(self.betting[0], 2)
                to_act = self.move_catcher.to_act#}}}
